### ISSUE TITLE
CLI for creating templates with scaffolder

### ISSUE THREAD

**Issue by timja** [backstage/backstage#13668]
## Feature Suggestion

Something like:
```
backstage-cli template --vars
```

## Context

I want to run tests against my project to make sure that it works after templating. Like how you could with the cookiecutter CLI.

Originally https://github.com/backstage/backstage/issues/6568, it was closed prematurely in my opinion.

There's two main usecases I see.

1. Iterating while developing a template, https://github.com/backstage/backstage/issues/6568 likely helped with that
2. Continuous integration of a template, no great solution that I can see, maybe https://github.com/jeremyben/nunjucks-cli or one of it's forks can be used in the meantime.

Also relates to https://github.com/backstage/backstage/issues/11853 but this is a specific feature rather than a broader issue

**Comment by benjdlambert** [comment]
Hmm, so we've looked at this before, and maybe the scope was a little too big. There was the idea of being able to run the entire template.yaml including custom actions from CLI, but that's a little tricky unfortunately, as you need all of the `customActions`, and `templateFilters`. Which is hard without having the whole source checked out too. So then we're left of the idea of calling a running Backstage instance and Scaffolder backend plugin but then I think it's going to be running untrusted template.yaml which can do anything so that's a big security implication and needs some thought.

If we were just to slim that down and basically offer a wrapper around nunjucks, we'd still need to have the ability to wrap up the `additionalTemplateFilters` somehow and be able to grab those which is kind of a similar problem to the whole custom actions part.

**Comment by timja** [comment]
go away bot

**Comment by fedyiv** [comment]
Hello!  We also need this feature, because if we can not build templates locally it would be impossible to have an automation to test them

**Comment by freben** [comment]
Noting briefly that this is partially blocked also by #15999

**Comment by timja** [comment]
seriously...

**Comment by benjdlambert** [comment]
@timja that should stop the stalebot :)

**Comment by lalitmathwani** [comment]
+1 - would make our end-to-end testing automation a lot simpler for one of our use cases at Adobe.

**Comment by gabrielucido** [comment]
+1 Would make our template development a lot faster!

**Comment by z-sourcecode** [comment]
+1 once you have a significant amount of templates it takes forever to re-create them manually on a fresh environment - automation is the key and necessary - security will come on top of it

**Comment by benjdlambert** [comment]
Yeah this is unfortunately still blocked by the upcoming work around authentication, specially CLI tokens that we can use to communicate with the running `scaffolder-backend` API to create jobs.

Should hopefully get around to starting some of this auth work early in the new year 🤞

**Comment by johanneswuerbach** [comment]
I'm wondering why this needs authentication as from my point of view an ideal workflow wouldn't require a running backstage instance.

Ideally in the CI of a repository with templates I could just do a `backstage-cli template --vars a=b` to render the templates  and run some checks on the output afterwards.

**Comment by benjdlambert** [comment]
@johanneswuerbach all the actions are defined in the `scaffolder-backend`. What you're asking for is basically a wrapper around the `fetch:template` action, which of course maybe useful on it's own, but if you're also defining things like `additionalFilters` they're not going to be available in the CLI without hitting a deployed instance. 

Effectively, this CLI would `dryRun` a template for you, and then get you back the workspace output. So it solves a much broader use case than just `fetch:template`. It's going to run templates end to end, and be more respective of a real template rather than just an isolated bit. Does this make sense?

**Comment by bobo** [comment]
We solved this at volvo-cars by adding a `yarn generate [template-name]` command to our central software-templates repository.
Each template has a "test-data.json" file that is used when generating the template, to make it easier to run.
Each template also have a github action that renders the template and runs the application tests, ie for java it will run `mvn test` on each PR.

If this sounds like a good aproach, we could try and get our generate script into backstage-cli? Or do you prefer to wait for the auth changes.

**Comment by awanlin** [comment]
Another +1 for this, would very much like to see this in place as it would greatly help testing Software Templates 👍

**Comment by mryan43** [comment]
> We solved this at volvo-cars by adding a `yarn generate [template-name]` command to our central software-templates repository. Each template has a "test-data.json" file that is used when generating the template, to make it easier to run. Each template also have a github action that renders the template and runs the application tests, ie for java it will run `mvn test` on each PR.
> 
> If this sounds like a good aproach, we could try and get our generate script into backstage-cli? Or do you prefer to wait for the auth changes.

do you share that code somewhere (generate [template-name]) ? the lack of a way to test templates is a pretty big blocker for us as well and such a workaround would really help us

**Comment by SaraJendeberg** [comment]
> > We solved this at volvo-cars by adding a `yarn generate [template-name]` command to our central software-templates repository. Each template has a "test-data.json" file that is used when generating the template, to make it easier to run. Each template also have a github action that renders the template and runs the application tests, ie for java it will run `mvn test` on each PR.
> > If this sounds like a good aproach, we could try and get our generate script into backstage-cli? Or do you prefer to wait for the auth changes.
> 
> do you share that code somewhere (generate [template-name]) ? the lack of a way to test templates is a pretty big blocker for us as well and such a workaround would really help us

@awanlin @mryan43 we have just started working on a `scaffolder-cli` that we hope to contribute to the project 😄 Will get back to you when we have something that you can try out!!

**Comment by awanlin** [comment]
Hi @SaraJendeberg, that's great! I highly recommend you attend one of the [Scaffolder SIG](https://github.com/backstage/community/tree/main/sigs/sig-scaffolder) meetings and speak with the Scaffolder maintainers to make sure your CLI aligns with their plans. I know the [CLI modularization BEP](https://github.com/backstage/backstage/pull/24904) is still in the PR phase and they have indicated that would be a blocker for any CLI work.

**Comment by SaraJendeberg** [comment]
> Hi @SaraJendeberg, that's great! I highly recommend you attend one of the [Scaffolder SIG](https://github.com/backstage/community/tree/main/sigs/sig-scaffolder) meetings and speak with the Scaffolder maintainers to make sure your CLI aligns with their plans. I know the [CLI modularization BEP](https://github.com/backstage/backstage/pull/24904) is still in the PR phase and they have indicated that would be a blocker for any CLI work.

Hi! @awanlin Thanks. Yes, it was mentioned in [discord](https://discord.com/channels/687207715902193673/1254696393478639677/1254737478464573440) when we brought up the idea and it was suggested to start on a `scaffolder-cli` separately and merge.

**Comment by awanlin** [comment]
Hi @SaraJendeberg, thanks for that link to the chat on Discord, that helps me out a lot. I know that a large portion of the Community would like this feature and I want to make sure it lands in a way that works well for everyone. Looks like Ben is good with this and he would know all the moving parts best. 🚀

**Comment by SaraJendeberg** [comment]
@Rugvip @benjdlambert @awanlin I have created a PR [here](https://github.com/backstage/backstage/issues/13668) that I think covers everything you mentioned except from the `scaffolder-cli login` part.
